Fixes #30206 - Targeting hosts - Search & pagination as React components

stejskalleos opened the pull request:
Requires https://github.com/theforeman/foreman/pull/7771

* Searching and pagination is now done in React way --> No page reloads
* Fixed issue with Loading spinner: when no items were found, loading spinner was displayed instead of message `No hosts found`
* `job_invocations_controller.rb#show` refactored, resources for Targeting hosts table are loaded only when they are needed.

ares commented on app/controllers/job_invocations_controller.rb:
```diff
@@ -52,16 +52,9 @@ def create
-    @hosts = resource_base_search_and_page
     @job_organization = Taxonomy.find_by(id: @job_invocation.task.input[:current_organization_id])
     @job_location = Taxonomy.find_by(id: @job_invocation.task.input[:current_location_id])
+    targeting_hosts_resources if params[:format] == 'json'
```
just a nit - could we just use `respond_to do |format|` [pattern](https://api.rubyonrails.org/classes/ActionController/MimeResponds.html#method-i-respond_to)?

MariaAga commented on webpack/react_app/components/TargetingHosts/index.js:
```diff
@@ -1,37 +1,99 @@
+
+  const handleSearch = query => {
+    const defaultPagination = { page: 1, perPage: pagination.perPage };
+    dispatch(stopInterval(TARGETING_HOSTS));
```
Should only run is the interval exists otherwise it throws an error that the interval doesnt exists 
You can select it using `foremanReact/redux/middlewares/IntervalMiddleware/IntervalSelectors.js`

stejskalleos replied:
I wasn't able reproduce the errors, but being 100% sure added new function
```
+  const stopApiInterval = () => {
+    if (intervalExists) {
+      dispatch(stopInterval(TARGETING_HOSTS));
+    }
+  };

MariaAga replied:
My bad, theforeman/foreman#7771 should fix this

MariaAga commented on webpack/react_app/components/TargetingHosts/index.js:
```diff
@@ -1,37 +1,99 @@
+  };
+
+  const handlePagination = args => {
+    dispatch(stopInterval(TARGETING_HOSTS));
```
Should only run is the interval exists otherwise it throws an error that the interval doesnt exists 
You can select it using `foremanReact/redux/middlewares/IntervalMiddleware/IntervalSelectors.js`

stejskalleos replied:
I wasn't able reproduce the errors, but being 100% sure added new function
```
+  const stopApiInterval = () => {
+    if (intervalExists) {
+      dispatch(stopInterval(TARGETING_HOSTS));
+    }
+  };

```

MariaAga commented on webpack/react_app/components/TargetingHosts/index.js:
```diff
@@ -1,37 +1,99 @@
 
   useEffect(() => {
-    dispatch(getData());
+    dispatch(getData(apiUrl));
```
When apiUrl changes the interval is called again without being stopped first and causes an error

stejskalleos replied:
Really? Because I don't get any errors on my setup. Do you have deployed https://github.com/theforeman/foreman/pull/7771 on Foreman? Without it my PR doesn't work.

MariaAga replied:
Now it work, thanks for the clarification

stejskalleos commented on webpack/react_app/components/TargetingHosts/index.js:
```diff
@@ -1,37 +1,99 @@
 
-  return <TargetingHosts status={status} items={items} />;
+  return (
+    <div id="targeting_hosts">
+      <div className="row">
+        <div className="title_filter col-md-6">
+          <SearchBar
+            onSearch={query => handleSearch(query)}
+            data={{
+              autocomplete: {
+                id: 'targeting_hosts_search',
+                url: '/hosts/auto_complete_search',
+              },
+            }}
+          />
+        </div>
+      </div>
+      <br />
+      <TargetingHosts status={status} items={items} autoRefresh={autoRefresh} />
+      <Pagination
+        viewType="list"
+        itemCount={totalHosts}
+        pagination={pagination}
+        onChange={args => handlePagination(args)}
+        dropdownButtonId="targeting-hosts-pagination-dropdown"
+      />
+    </div>
```
You mean to rename `index.js` to `TargetingHostsPage.js`, or create `TargetingHostsPage` component and render it in 'index.js'?

MariaAga replied:
The second one, render `TargetingHostsPage` in `index.js`

stejskalleos commented on webpack/react_app/components/TargetingHosts/index.js:
```diff
@@ -1,37 +1,99 @@
 
   useEffect(() => {
-    dispatch(getData());
+    dispatch(getData(apiUrl));
+
+    if (autoRefresh === 'false') {
+      dispatch(stopInterval(TARGETING_HOSTS));
+    }
```
It must be there, it prevents never-ending loop of API calls after pagination change.

MariaAga replied:
So can you remove the second useEffect?

MariaAga commented on webpack/react_app/components/TargetingHosts/TargetingHostsPage.js:
```diff
@@ -0,0 +1,63 @@
+              url: '/hosts/auto_complete_search',
+              useKeyShortcuts: true,
+            },
+            bookmarks: {},
```
We don't want bookmarks in this search?

stejskalleos replied:
Not in this PR, because if you enable bookmarks they reload the page with new URL, but `TargetingHosts` component ignores params from URL so nothing will change, just page reload.

Since they weren't there before I don't think it's a big issue.

MariaAga replied:
Yeah makes sense

MariaAga commented on webpack/react_app/components/TargetingHosts/TargetingHosts.js:
```diff
@@ -17,7 +17,7 @@ const TargetingHosts = ({ status, items }) => {
 
   return (
-    <LoadingState loading={!items.length}>
+    <LoadingState loading={!items.length && autoRefresh === 'true'}>
```
Why not check here if the status is pending and no items?

MariaAga commented on webpack/react_app/components/TargetingHosts/TargetingHostsHelpers.js:
```diff
@@ -0,0 +1,9 @@
+export const getApiUrl = (searchQuery, pagination) =>
+  getURI()
+    .search('')
+    .addQuery('search', searchQuery)
```
Empty search eg :`/job_invocations/15?search=&page=1&per_page=20&format=json` causes the hosts list to be empty for queued jobs for me, can you not add the search if its empty or maybe fix it in the controller?

stejskalleos commented:
Rebased & updated.

* Fixed bug with never-ending loop when `apiStatus === STATUS.ERROR`
* Fixed bug with empty searchQuery and queued job.
* Some minor refactorings ...

MariaAga commented on webpack/react_app/components/TargetingHosts/__tests__/__snapshots__/TargetingHosts.test.js.snap:
```diff
@@ -52,7 +52,7 @@ exports[`TargetingHosts renders with error 1`] = `
 <LoadingState
   additionalClasses=""
-  loading={true}
+  loading={false}
```
The title says `TargetingHosts renders with loading 1` so this should be true

adamruzicka commented:
Since theforeman/foreman#7771 will be available only for Foreman >=2.2, could you bump the required foreman version in `engine.rb`?

laviro commented on webpack/react_app/components/TargetingHosts/TargetingHostsActions.js:
```diff
@@ -1,8 +1,6 @@
 
-const url = getURI().addQuery('format', 'json');
-export const getData = () =>
+export const getData = url =>
```
I think we can move this method to the `index.js` file? since it's already using the middleware helpers.
after that you can also delete this file :)

stejskalleos replied:
Can I keep it there? I have another PR where I extend this file with more functions.

laviro replied:
forget what I said then :)

laviro commented on webpack/react_app/components/TargetingHosts/TargetingHostsPage.js:
```diff
@@ -0,0 +1,61 @@
+  handlePagination,
+}) => (
+  <div id="targeting_hosts">
+    <div className="row">
```
can you use `Grid.Row` component from Patternfly?

laviro commented on webpack/react_app/components/TargetingHosts/TargetingHostsPage.js:
```diff
@@ -0,0 +1,61 @@
+}) => (
+  <div id="targeting_hosts">
+    <div className="row">
+      <div className="title_filter col-md-6">
```
Can you use `<Grid.Col md={6}>` here?

stejskalleos commented:
> Awesome work @stejskalleos ! works great :)
> left some suggestions, can you also add tests for:
> 
> * TargetingHosts.js
> * TargetingHostsHelpers.js
> * TargetingHostsPage.js
> * TargetingHostsSelectors.js

Updated & rebased

laviro commented on app/views/job_invocations/_tab_hosts.html.erb:
```diff
@@ -1,24 +1,5 @@
-    <%= mount_react_component('TargetingHosts', '#targeting_hosts') %>
-  </div>
-  <%= will_paginate_with_info @hosts, :container => true %>
+  <%= react_component('TargetingHosts' ) %>
```
it will be really cool if we could move all of this login into the client side,
passing the `job_invocation` as a `prop`, something like: 
`<%= react_component('TargetingHosts' , { job_invocation }) %>`

so the `if` and `else` logic will be implemented in the component already :)

stejskalleos replied:
Can we do it in another PR? This PR is already big and I don't want to add additional changes.

laviro replied:
sure, though it should be something like 5-10 lines changes :pray:

laviro commented on webpack/react_app/components/TargetingHosts/__tests__/__snapshots__/TargetingHosts.test.js.snap:
```diff
@@ -74,7 +74,15 @@ exports[`TargetingHosts renders with loading 1`] = `
+          <td
+            colSpan="3"
+          >
+            No hosts found.
```
interesting, in the fixtures there is one item

stejskalleos replied:
It's correct, for loading scenario `hosts array` is empty:
```
  'renders with loading': {
    apiStatus: 'PENDING',
    items: [],
  },
```

laviro commented on webpack/react_app/components/TargetingHosts/__tests__/fixtures.js:
```diff
@@ -37,7 +37,29 @@ export const TargetingHostsFixtures = {
+    handleSearch: () => {},
+    searchQuery: '',
+    apiStatus: 'RESOLVED',
+    items: [
+      {
+        name: 'host',
+        link: '/link',
+        status: 'success',
+        actions: [],
+      },
+    ],
```
I think you can re-use this object in all of the fixtures, just declare it on the top of the page.

laviro commented on webpack/react_app/components/TargetingHosts/TargetingHostsPage.scss:
```diff
@@ -0,0 +1,3 @@
+.targeting-hosts-pagination {
+  margin-top: -7px;
+}
```
can you also add `min-height: 350px` to the page wrapper?
so it will be easy to see the table, and it won't jump when the searchbar opens:

![Selection_335](https://user-images.githubusercontent.com/26363699/86786061-32d45a80-c06c-11ea-950a-ef726cf504ec.png)

page ends here :arrow_up:

laviro commented on webpack/react_app/components/TargetingHosts/index.js:
```diff
@@ -1,37 +1,87 @@
+  }, [dispatch, apiUrl, autoRefresh]);
+
+  if (apiStatus === STATUS.ERROR) {
+    stopApiInterval();
```
you can use the new api middleware callbacks, instead of waiting for the state to update with the status
e.g: 
```js
get({ 
  key: TARGETING_HOSTS, 
  url,
  handleError: () => { stopInterval() } 
});
```

stejskalleos replied:
It doesn't work for me, api call end up in never-ending loop.

laviro replied:
just make sure that you pass it with `dispatch` :)

adamruzicka commented:
I'm getting this when clicking on next page
![image](https://user-images.githubusercontent.com/7326770/87025982-6255a500-c1db-11ea-9d67-50570c3e36bb.png)

MariaAga commented:
I tried moving between pages and didn't get any error, weird..

stejskalleos commented:
> I'm getting this when clicking on next page

Do you have latest `develop` branch in foreman-core? I don't get any errors when changing pages.

adamruzicka commented:
Ah, right. Didn't have the latest and greatest. Shouldn't there be some kind of spinner or loading indicator when new data is being loaded?

stejskalleos commented:
There is a spinner, when table is loading for the first time. But no spinner when you change the page (same behavior as other tables in foreman)

adamruzicka commented:
It is kinda odd when you click a button and nothing happens for a few seconds and then things change, but if that's how the tables behave everywhere let's stay consistent

MariaAga commented:
I think in tasks we added a spinner next to the search bar to show searching if you want to add it in another PR

adamruzicka commented:
Alright, let's keep it for later. One last thing, could you drop 2.1-stable from the gh action definition since this is develop/2.2+ only?

stejskalleos commented:
> Alright, let's keep it for later. One last thing, could you drop 2.1-stable from the gh action definition since this is develop/2.2+ only?

Done & green.
